Adapt invocation export UI for plugins

davelopez opened the pull request:
This PR unifies the workflow invocation export UI while providing an easy way of defining new export formats. Every format can be downloaded or written to a remote source by default. Each export plugin can define additional custom operations depending on the particular needs.

## Export invocation to RO-crate
![export-ro-crate-invocation-ui](https://user-images.githubusercontent.com/46503462/191081589-c67901b5-4a5a-4aa3-b2e4-1b22c402729b.gif)

## Export invocation to BioCompute Object (requires #14620)
In addition to the direct download and export to a remote source, this export plugin allows sending a BCO to an external database.

![send-bco-form-ui](https://user-images.githubusercontent.com/46503462/191082400-397cbd81-0a8f-41f8-a2c5-a604121ead76.gif)

## How to add new export plugins
Once the backend supports a new invocation export format, we can add a new `InvocationExportPlugin` instance inside `client/src/components/Workflow/Invocation/Export/Plugins/` and register it in the *index.js* file. And that's it!

For the sake of consistency, If the plugin needs to display additional components as part of the custom actions, it should be done inside a modal like in the example above, and the additional component should be placed in a subfolder with the name of the plugin.


## TODO
- [x] Write some tests

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

jmchilton committed:
Add post parameters to StsDownloadButton...

jmchilton committed:
UI for Workflow Export.

davelopez committed:
Move StsDownloadButton to components

It is generic enough to be used outside the markdown component

davelopez committed:
Add workflow invocation export plugins

Just a basic structure to define an invocation export plugin

davelopez committed:
Make StsDownloadButton variant optional

davelopez committed:
Unify invocation export UI for plugins

Using composition API

nsoranzo commented on client/src/components/Workflow/InvocationExport/InvocationExport.vue:
```diff
@@ -0,0 +1,55 @@
+<template>
+    <span class="workflow-invocation-export-component">
+        <h2>Export worklfow invocation</h2>
```
```suggestion
        <h2>Export workflow invocation</h2>
```

davelopez replied:
Thanks Nicola for the catch, but I'm considering dropping this component in favor of the plugin system. I want to discuss it with John first :)

davelopez committed:
Add RO-crate invocation export plugin

davelopez committed:
Move markdown rendered to composable

davelopez committed:
Add file sources composable and modal dialog

davelopez committed:
Drop unused InvocationExport component

It was provisional

davelopez committed:
Add useTaskMonitor composable

This composable simplifies polling to wait for a particular background task to complete.

davelopez committed:
Pass download parameters to STS download button

davelopez committed:
Remove unused useFilesSources

Is a YAGNI for now...

davelopez committed:
Add support for remote file sources export

davelopez committed:
Add export action button component

davelopez committed:
Move fontawesome imports out of script setup

davelopez committed:
Add BCO send form as additional export action

davelopez committed:
Display success or error feedback after export

davelopez committed:
Refactor useMarkdown composable + add unit tests

davelopez committed:
Add unit tests for useTaskMonitor composable

davelopez committed:
Use correct download format for default file export

davelopez committed:
Add some docstrings

davelopez committed:
Cleanup unused client route

davelopez committed:
Add some tests for InvocationExportPluginCard component

davelopez committed:
Update default prefix to `GALXY`, a publicly available prefix to group Galaxy generated BCOs.

davelopez committed:
Move BioComputeObject plugin to own directory

davelopez committed:
Generate asynchronously BCO to send to DB

itisAliRH commented on client/src/components/Workflow/Invocation/Export/ExportToRemoteButton.vue:
```diff
@@ -0,0 +1,51 @@
+</script>
+
+<template>
+    <div>
```
The group button UI broke because of this parent `div` and can be deleted.
Current (look at the borders)
![image](https://user-images.githubusercontent.com/8046843/196416214-70f8edf2-84a4-481c-ad78-a21d70f44fda.png)
![image](https://user-images.githubusercontent.com/8046843/196415809-079ed399-6d32-46ce-b655-184407776e95.png)

Correct:
![image](https://user-images.githubusercontent.com/8046843/196415951-ea2ea371-79be-4642-a547-5a20c84561eb.png)
![image](https://user-images.githubusercontent.com/8046843/196415922-fda19700-890d-409b-9100-7711c6cde9ea.png)

davelopez replied:
Ohh nice catch! Thank you!

itisAliRH commented on client/src/components/Workflow/Invocation/Export/ActionButton.vue:
```diff
@@ -0,0 +1,22 @@
+
+<template>
+    <b-button v-b-tooltip.hover.bottom :title="props.action.title" @click="props.action.run(modal)">
+        <font-awesome-icon v-if="props.action.icon" :icon="props.action.icon" />
```
```suggestion
        <icon v-if="props.action.icon" :icon="props.action.icon" />
```
The `icon` component is registered globally for font awesome icons and it can be replaced here and in other files too :)

dannon replied:
I'm torn on `icon` and I don't think we need to use it across the board.  It's okay to be explicit and use the font-awesome-icon directly.

davelopez replied:
Ok, leaving it as `font-awesome-icon` then!

davelopez committed:
Address comments from code review

- Remove div left over.
- Set autocomplete in form inputs to `off` instead of invalid value.

itisAliRH committed:
Add alert to improve UX for workflow load, summary and state
